File: scripts/viva_network.py
#!/usr/bin/env python3
"""
VIVA Network Science Visualization
Analyzes holographic memory space dynamics using NetworkX

Usage: python scripts/viva_network.py
"""

import os
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import networkx as nx
from scipy.spatial.distance import pdist, squareform

# Configuration
VIVA_API = "http://localhost:8888"
OUTPUT_DIR = "output"
PROXIMITY_THRESHOLD = 5.0  # Link bodies within this distance

# Ensure output directory exists
os.makedirs(OUTPUT_DIR, exist_ok=True)

print("=== VIVA Network Science Visualization (Python) ===\n")

# Fetch data
print(f"Fetching data from {VIVA_API}...")
try:
    df = pd.read_csv(f"{VIVA_API}/api/export/csv")
except Exception as e:
    print(f"Error: Could not connect to VIVA server.")
    print(f"Make sure the server is running: gleam run -m viva/telemetry/demo")
    sys.exit(1)

print(f"Loaded {len(df)} memory bodies\n")

# === Build NetworkX Graph ===
print("Building network graph...")
G = nx.Graph()

# Add nodes
for _, row in df.iterrows():
    G.add_node(
        row['id'],
        label=row['label'],
        pos=(row['x'], row['y']),
        pos_zw=(row['z'], row['w']),
        energy=row['energy'],
        sleeping=row['sleeping'],
        island_id=row['island_id']
    )

# Add edges based on proximity in HRR-4D space
coords = df[['x', 'y', 'z', 'w']].values
distances = squareform(pdist(coords))

for i in range(len(df)):
    for j in range(i + 1, len(df)):
        if distances[i, j] < PROXIMITY_THRESHOLD:
            G.add_edge(
                df.iloc[i]['id'],
                df.iloc[j]['id'],
                weight=1.0 / (1.0 + distances[i, j])
            )

print(f"  Nodes: {G.number_of_nodes()}")
print(f"  Edges: {G.number_of_edges()}")

# === Network Metrics ===
print("\n=== Network Metrics ===")

# Density
density = nx.density(G)
print(f"Density (ρ): {density:.4f}")

# Degree distribution
degrees = [d for n, d in G.degree()]
avg_degree = np.mean(degrees) if degrees else 0
print(f"Average degree (k̄): {avg_degree:.2f}")

# Clustering coefficient
clustering = nx.average_clustering(G)
print(f"Clustering coefficient (C): {clustering:.4f}")

# Connected components
components = nx.number_connected_components(G)
print(f"Connected components: {components}")

# === 1. Network Graph Visualization ===
print("\nGenerating network visualization...")
fig, ax = plt.subplots(figsize=(12, 10))

# Node positions from HRR coordinates
pos = {row['id']: (row['x'], row['y']) for _, row in df.iterrows()}

# Node colors based on energy
energies = [G.nodes[n]['energy'] for n in G.nodes()]
node_colors = plt.cm.plasma(np.array(energies) / max(energies) if max(energies) > 0 else energies)

# Node sizes based on energy
node_sizes = [300 + e * 1000 for e in energies]

# Draw network
nx.draw_networkx_edges(G, pos, alpha=0.3, width=1, edge_color='gray', ax=ax)
nx.draw_networkx_nodes(G, pos, node_color=energies, cmap=plt.cm.plasma,
                       node_size=node_sizes, alpha=0.8, ax=ax)

# Labels
labels = {row['id']: row['label'] for _, row in df.iterrows()}
nx.draw_networkx_labels(G, pos, labels, font_size=9, font_color='black', ax=ax)

ax.set_title("VIVA Memory Network (HRR-4D → 2D)", fontsize=16, fontweight='bold')
ax.set_xlabel("X (HRR dimension 1)")
ax.set_ylabel("Y (HRR dimension 2)")
ax.axis('equal')

# Colorbar
sm = plt.cm.ScalarMappable(cmap=plt.cm.plasma, norm=plt.Normalize(vmin=min(energies), vmax=max(energies)))
sm.set_array([])
cbar = plt.colorbar(sm, ax=ax, shrink=0.8)
cbar.set_label('Energy')

plt.tight_layout()
plt.savefig(os.path.join(OUTPUT_DIR, "viva_network.png"), dpi=150)
print(f"  -> {OUTPUT_DIR}/viva_network.png")
plt.close()

# === 2. Degree Distribution ===
print("Generating degree distribution...")
fig, ax = plt.subplots(figsize=(10, 6))

degree_counts = pd.Series(degrees).value_counts().sort_index()
ax.bar(degree_counts.index, degree_counts.values, color='steelblue', alpha=0.8, edgecolor='white')
ax.axvline(avg_degree, color='coral', linestyle='--', linewidth=2, label=f'Mean = {avg_degree:.2f}')

ax.set_xlabel("Degree (k)")
ax.set_ylabel("Count")
ax.set_title("Degree Distribution", fontsize=14, fontweight='bold')
ax.legend()

plt.tight_layout()
plt.savefig(os.path.join(OUTPUT_DIR, "viva_degree_dist.png"), dpi=150)
print(f"  -> {OUTPUT_DIR}/viva_degree_dist.png")
plt.close()

# === 3. Distance Matrix Heatmap ===
print("Generating distance matrix heatmap...")
fig, ax = plt.subplots(figsize=(10, 8))

labels_list = df['label'].tolist()
im = ax.imshow(distances, cmap='viridis_r', aspect='equal')

ax.set_xticks(range(len(labels_list)))
ax.set_yticks(range(len(labels_list)))
ax.set_xticklabels(labels_list, rotation=45, ha='right')
ax.set_yticklabels(labels_list)

ax.set_title("HRR-4D Distance Matrix", fontsize=14, fontweight='bold')
cbar = plt.colorbar(im, ax=ax, shrink=0.8)
cbar.set_label('Euclidean Distance')

plt.tight_layout()
plt.savefig(os.path.join(OUTPUT_DIR, "viva_distance_matrix.png"), dpi=150)
print(f"  -> {OUTPUT_DIR}/viva_distance_matrix.png")
plt.close()

# === 4. Energy vs Position ===
print("Generating energy-position plot...")
fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# X vs Energy
axes[0, 0].scatter(df['x'], df['energy'], c=df['energy'], cmap='plasma', s=100, alpha=0.8)
axes[0, 0].set_xlabel('X')
axes[0, 0].set_ylabel('Energy')
axes[0, 0].set_title('Energy vs X')

# Y vs Energy
axes[0, 1].scatter(df['y'], df['energy'], c=df['energy'], cmap='plasma', s=100, alpha=0.8)
axes[0, 1].set_xlabel('Y')
axes[0, 1].set_ylabel('Energy')
axes[0, 1].set_title('Energy vs Y')

# Z vs Energy
axes[1, 0].scatter(df['z'], df['energy'], c=df['energy'], cmap='plasma', s=100, alpha=0.8)
axes[1, 0].set_xlabel('Z')
axes[1, 0].set_ylabel('Energy')
axes[1, 0].set_title('Energy vs Z')

# W vs Energy
axes[1, 1].scatter(df['w'], df['energy'], c=df['energy'], cmap='plasma', s=100, alpha=0.8)
axes[1, 1].set_xlabel('W')
axes[1, 1].set_ylabel('Energy')
axes[1, 1].set_title('Energy vs W')

plt.suptitle("Energy Distribution Across HRR-4D Dimensions", fontsize=14, fontweight='bold')
plt.tight_layout()
plt.savefig(os.path.join(OUTPUT_DIR, "viva_energy_position.png"), dpi=150)
print(f"  -> {OUTPUT_DIR}/viva_energy_position.png")
plt.close()

# === Summary ===
print("\n=== Summary ===")
print(f"Total bodies: {len(df)}")
print(f"Active: {(~df['sleeping']).sum()}")
print(f"Dormant: {df['sleeping'].sum()}")
print(f"Islands: {df['island_id'].nunique()}")
print(f"Mean energy: {df['energy'].mean():.4f}")
print(f"Energy variance: {df['energy'].var():.6f}")

print("\n=== Done! ===")

File: scripts/viva_network.R
#!/usr/bin/env Rscript
# VIVA Network Science Visualization
# Analyzes holographic memory space dynamics
#
# Usage: Rscript scripts/viva_network.R

library(tidyverse)
library(viridis)

# Configuration
VIVA_API <- "http://localhost:8888"
OUTPUT_DIR <- "output"

# Ensure output directory exists
dir.create(OUTPUT_DIR, showWarnings = FALSE)

cat("=== VIVA Network Science Visualization ===\n\n")

# Fetch data from VIVA telemetry server
cat("Fetching data from", VIVA_API, "...\n")
data <- tryCatch({
  read_csv(paste0(VIVA_API, "/api/export/csv"), show_col_types = FALSE)
}, error = function(e) {
  cat("Error: Could not connect to VIVA server.\n")
  cat("Make sure the server is running: gleam run -m viva/telemetry/demo\n")
  quit(status = 1)
})

cat("Loaded", nrow(data), "memory bodies\n\n")

# === 1. HRR-4D Scatter Plot (x, y projection) ===
cat("Generating HRR-4D scatter plot...\n")
p1 <- ggplot(data, aes(x = x, y = y, color = energy, size = energy)) +
  geom_point(alpha = 0.8) +
  geom_text(aes(label = label), vjust = -1.5, size = 3, color = "gray30") +
  scale_color_viridis(option = "plasma", name = "Energy") +
  scale_size_continuous(range = c(3, 12), guide = "none") +
  labs(
    title = "VIVA Holographic Memory Space",
    subtitle = "HRR-4D projected to 2D (x, y coordinates)",
    x = "X (HRR dimension 1)",
    y = "Y (HRR dimension 2)"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(color = "gray50"),
    panel.grid.minor = element_blank()
  ) +
  coord_fixed()

ggsave(file.path(OUTPUT_DIR, "viva_scatter_xy.png"), p1, width = 10, height = 8, dpi = 150)
cat("  -> output/viva_scatter_xy.png\n")

# === 2. Alternative projection (z, w) ===
cat("Generating z-w projection...\n")
p2 <- ggplot(data, aes(x = z, y = w, color = energy, size = energy)) +
  geom_point(alpha = 0.8) +
  geom_text(aes(label = label), vjust = -1.5, size = 3, color = "gray30") +
  scale_color_viridis(option = "magma", name = "Energy") +
  scale_size_continuous(range = c(3, 12), guide = "none") +
  labs(
    title = "VIVA Memory Space (Z-W Projection)",
    subtitle = "Higher dimensions of HRR-4D space",
    x = "Z (HRR dimension 3)",
    y = "W (HRR dimension 4)"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(color = "gray50")
  ) +
  coord_fixed()

ggsave(file.path(OUTPUT_DIR, "viva_scatter_zw.png"), p2, width = 10, height = 8, dpi = 150)
cat("  -> output/viva_scatter_zw.png\n")

# === 3. Energy Distribution ===
cat("Generating energy distribution...\n")
p3 <- ggplot(data, aes(x = energy, fill = sleeping)) +
  geom_histogram(bins = 20, alpha = 0.8, color = "white") +
  geom_density(aes(y = after_stat(count)), alpha = 0.3, fill = NA, color = "purple", linewidth = 1) +
  scale_fill_manual(values = c("TRUE" = "steelblue", "FALSE" = "coral"),
                    labels = c("TRUE" = "Dormant", "FALSE" = "Active"),
                    name = "State") +
  labs(
    title = "Energy Distribution of Memory Bodies",
    subtitle = paste("N =", nrow(data), "bodies"),
    x = "Energy Level",
    y = "Count"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(color = "gray50")
  )

ggsave(file.path(OUTPUT_DIR, "viva_energy_dist.png"), p3, width = 10, height = 6, dpi = 150)
cat("  -> output/viva_energy_dist.png\n")

# === 4. Island Clustering ===
cat("Generating island clustering...\n")
p4 <- ggplot(data, aes(x = x, y = y, color = factor(island_id))) +
  geom_point(size = 5, alpha = 0.8) +
  geom_text(aes(label = label), vjust = -1.5, size = 3, color = "gray30") +
  scale_color_brewer(palette = "Set2", name = "Island") +
  labs(
    title = "Memory Clustering by Island",
    subtitle = "Spatial grouping of related memories",
    x = "X", y = "Y"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(color = "gray50")
  ) +
  coord_fixed()

ggsave(file.path(OUTPUT_DIR, "viva_islands.png"), p4, width = 10, height = 8, dpi = 150)
cat("  -> output/viva_islands.png\n")

# === 5. Summary Statistics ===
cat("\n=== Summary Statistics ===\n")
cat("Total bodies:", nrow(data), "\n")
cat("Active:", sum(!data$sleeping), "\n")
cat("Dormant:", sum(data$sleeping), "\n")
cat("Islands:", n_distinct(data$island_id), "\n")
cat("Mean energy:", round(mean(data$energy), 4), "\n")
cat("Energy variance:", round(var(data$energy), 6), "\n")

# Spatial spread
cat("\nSpatial extent:\n")
cat("  X range: [", round(min(data$x), 2), ",", round(max(data$x), 2), "]\n")
cat("  Y range: [", round(min(data$y), 2), ",", round(max(data$y), 2), "]\n")
cat("  Z range: [", round(min(data$z), 2), ",", round(max(data$z), 2), "]\n")
cat("  W range: [", round(min(data$w), 2), ",", round(max(data$w), 2), "]\n")

cat("\n=== Done! ===\n")
